tests/unit/s2n_client_key_share_extension_test.c

```diff
@@ -393,6 +394,42 @@ int main(int argc, char **argv)
         }
     }
 
+    {
+        /* Test that unknown TLS_EXTENSION_SUPPORTED_GROUPS values are ignored */
+        {
+            struct s2n_ecc_named_curve unsupported_curves[2] = {
+                    { .iana_id = 0x0, .libcrypto_nid = 0, .name = 0x0, .share_size = 0 },
+                    { .iana_id = 0xFF01, .libcrypto_nid = 0, .name = 0x0, .share_size = 0 },
+            };
+            int ec_curves_count = s2n_array_len(unsupported_curves);
+            struct s2n_connection *conn;
+            EXPECT_NOT_NULL(conn = s2n_connection_new(S2N_SERVER));
+
+            struct s2n_stuffer supported_groups_extension;
+            EXPECT_SUCCESS(s2n_stuffer_alloc(&supported_groups_extension, 2 + ec_curves_count * 2));
+            GUARD(s2n_stuffer_write_uint16(&supported_groups_extension, ec_curves_count * 2));
+            for (int i = 0; i < ec_curves_count; i++) {
+                GUARD(s2n_stuffer_write_uint16(&supported_groups_extension, unsupported_curves[i].iana_id));
+            }
+
+            /* Create a properly parsed client hello extension using the stuffer's blob */
+            struct s2n_array *parsed_extensions = s2n_array_new(sizeof(struct s2n_client_hello_parsed_extension));
+            struct s2n_client_hello_parsed_extension *parsed_named_group_extension = s2n_array_pushback(parsed_extensions);
+            parsed_named_group_extension->extension_type = TLS_EXTENSION_SUPPORTED_GROUPS;
+            parsed_named_group_extension->extension = supported_groups_extension.blob;
+
+            /* Force a bad value for the negotiated curve so we know extension was parsed and the curve was set to NULL */
+            struct s2n_ecc_named_curve invalid_curve = { 0 };
+            conn->secure.server_ecc_params.negotiated_curve = &invalid_curve;
+            EXPECT_SUCCESS(s2n_client_extensions_recv(conn, parsed_extensions));
```

Comment:
Why use s2n_client_extensions_recv instead of s2n_extensions_client_key_share_recv?

Reply:
I could have used s2n_recv_client_supported_groups() and done the additional extension parsing in the unit test. I used s2n_client_extensions_recv() because it seemed like the canonical way to process a received extension.

This test is in s2n_client_key_share_extension_test because it seemed closely related, without having to create a new test file. Since this doesn't actually test the key share extension, maybe a new file is the way to go?

Reply:
Oh I see, the supported group extension is a tls1.2 extension so was written in the old style. You used s2n_client_extensions_recv because s2n_recv_client_supported_group is just a static method in the s2n_client_extensions file. 

Yes, this test is really unrelated to the key share extension. You could put it in s2n_client_extensions_test.c instead, but I'd suggest creating a separate file for it since that's the direction we're heading (tls1.3 extensions are all in separate files, with their own sets of unit tests).